### PR TITLE
fix: flake in teardown of test_unknown_dfly_env

### LINKED ISSUE #3314: generic_test.py::test_unknown_dfly_env
Fix the failure https://github.com/dragonflydb/dragonfly/actions/runs/9909935256/job/27379172007#step:6:723

### PR DESCRIPTION
**Pull request by kostasrim**
For reasons I do not exactly comprehend, `test_unknown_dfly_env` might fail when we try to `stop the instance`. The thing is, that dragonfly is already killed since we try to launch an instance with an invalid environment flag and `stop_all` should not actually have any effect. Since the error happens at teardown and when we call `stop_all` it makes sense to `skip it` since the instance is not alive in the first place. The fix is simple, allow skipping `stop_all` at instance teardown.

fixes #3314

* allow to skip stop_all on DflyInstanceFactory

### CODE REVIEW

**Comment by kostasrim** [comment]
For context (copy pasted the logs here):

```
2024-07-12T15:22:44.5047858Z [31m[1m___________ ERROR at teardown of test_unknown_dfly_env[df_factory0] ____________[0m
2024-07-12T15:22:44.5048701Z 
2024-07-12T15:22:44.5049409Z request = <SubRequest 'df_factory' for <Function test_unknown_dfly_env[df_factory0]>>
2024-07-12T15:22:44.5050590Z tmp_dir = PosixPath('/tmp/tmp2ogpm9xj')
2024-07-12T15:22:44.5052450Z test_env = {'CI': 'true', 'DEBIAN_FRONTEND': 'noninteractive', 'DRAGONFLY_PATH': '/__w/dragonfly/dragonfly/build/dragonfly', 'DRAGONFLY_TMP': '/tmp/tmp2ogpm9xj', ...}
2024-07-12T15:22:44.5054656Z 
2024-07-12T15:22:44.5054986Z     @pytest.fixture(scope="function", params=[{}])
2024-07-12T15:22:44.5056135Z     def df_factory(request, tmp_dir, test_env) -> DflyInstanceFactory:
2024-07-12T15:22:44.5057134Z         """
2024-07-12T15:22:44.5058881Z         Create an instance factory with supplied params.
2024-07-12T15:22:44.5059667Z         """
2024-07-12T15:22:44.5060388Z         os.makedirs(os.path.join(gettempdir(), "tiered"), exist_ok=True)
2024-07-12T15:22:44.5061469Z         scripts_dir = os.path.dirname(os.path.abspath(__file__))
2024-07-12T15:22:44.5062935Z         path = os.environ.get("DRAGONFLY_PATH", os.path.join(scripts_dir, "../../build-dbg/dragonfly"))
2024-07-12T15:22:44.5064034Z     
2024-07-12T15:22:44.5064605Z         log_directory = getattr(request.node, "log_dir")
2024-07-12T15:22:44.5065342Z     
2024-07-12T15:22:44.5065915Z         args = request.param if request.param else {}
2024-07-12T15:22:44.5066948Z         existing = request.config.getoption("--existing-port")
2024-07-12T15:22:44.5068304Z         existing_admin = request.config.getoption("--existing-admin-port")
2024-07-12T15:22:44.5069642Z         existing_mc = request.config.getoption("--existing-mc-port")
2024-07-12T15:22:44.5070529Z         params = DflyParams(
2024-07-12T15:22:44.5071123Z             path=path,
2024-07-12T15:22:44.5071662Z             cwd=tmp_dir,
2024-07-12T15:22:44.5072416Z             gdb=request.config.getoption("--gdb"),
2024-07-12T15:22:44.5073476Z             buffered_out=request.config.getoption("--buffered-output"),
2024-07-12T15:22:44.5074594Z             args=parse_args(request.config.getoption("--df")),
2024-07-12T15:22:44.5075535Z             existing_port=int(existing) if existing else None,
2024-07-12T15:22:44.5076594Z             existing_admin_port=int(existing_admin) if existing_admin else None,
2024-07-12T15:22:44.5081039Z             existing_mc_port=int(existing_mc) if existing_mc else None,
2024-07-12T15:22:44.5081913Z             env=test_env,
2024-07-12T15:22:44.5082499Z             log_dir=log_directory,
2024-07-12T15:22:44.5083109Z         )
2024-07-12T15:22:44.5083547Z     
2024-07-12T15:22:44.5084105Z         factory = DflyInstanceFactory(params, args)
2024-07-12T15:22:44.5084855Z         yield factory
2024-07-12T15:22:44.5085390Z >       factory.stop_all()
2024-07-12T15:22:44.5085773Z 
2024-07-12T15:22:44.5086176Z [1m[31mdragonfly/conftest.py[0m:144: 
2024-07-12T15:22:44.5087171Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2024-07-12T15:22:44.5088261Z [1m[31mdragonfly/instance.py[0m:360: in stop_all
2024-07-12T15:22:44.5089025Z     instance.stop()
2024-07-12T15:22:44.5089722Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2024-07-12T15:22:44.5090372Z 
2024-07-12T15:22:44.5091264Z self = <[RuntimeError('port is not available yet') raised in repr()] DflyInstance object at 0xe89701335100>
2024-07-12T15:22:44.5092471Z kill = False
2024-07-12T15:22:44.5092762Z 
2024-07-12T15:22:44.5092990Z     def stop(self, kill=False):
2024-07-12T15:22:44.5093668Z         proc, self.proc = self.proc, None
2024-07-12T15:22:44.5094366Z         if proc is None:
2024-07-12T15:22:44.5094907Z             return
2024-07-12T15:22:44.5095388Z     
2024-07-12T15:22:44.5095987Z         logging.debug(f"Stopping instance on {self._port}")
2024-07-12T15:22:44.5096774Z         try:
2024-07-12T15:22:44.5097473Z             if kill:
2024-07-12T15:22:44.5098339Z                 proc.kill()
2024-07-12T15:22:44.5098956Z             else:
2024-07-12T15:22:44.5099481Z                 proc.terminate()
2024-07-12T15:22:44.5100160Z                 proc.communicate(timeout=15)
2024-07-12T15:22:44.5101014Z                 # if the return code is 0 it means normal termination
2024-07-12T15:22:44.5102061Z                 # if the return code is negative it means termination by signal
2024-07-12T15:22:44.5103812Z                 # if the return code is positive it means abnormal exit
2024-07-12T15:22:44.5104689Z                 if proc.returncode != 0:
2024-07-12T15:22:44.5105571Z >                   raise Exception("Dragonfly did not terminate gracefully")
2024-07-12T15:22:44.5106838Z [1m[31mE                   Exception: Dragonfly did not terminate gracefully[0m
2024-07-12T15:22:44.5107653Z 

```

File: tests/dragonfly/instance.py [hunk 1]
```diff
@@ -370,6 +371,9 @@ def start_all(self, instances: List[DflyInstance]):
 
     def stop_all(self):
         """Stop all launched instances."""
+        if self.skip_stop == True:
```
Comment (dranikpg) [inline tests/dragonfly/instance.py:374]: 🤓

**Review by dranikpg — APPROVED** [review]
Yet if it doesn't stop for some reason, it'll keep running unnoticed

File: tests/dragonfly/instance.py (hunk 1 above)
Comment (kostasrim) [reply tests/dragonfly/instance.py:374]: lol you are right :nerd_face:

**Comment by kostasrim** [comment]
> Yet if it doesn't stop for some reason, it'll keep running unnoticed

It will fail because the test won't throw an exception and it will be cleaned up by docker. The problem though exists if it;s local. I will move this into a function such that we skip only when we know that we did not throw an exception. I will push a fix

File: tests/dragonfly/generic_test.py [hunk 2]
```diff
@@ -100,9 +100,12 @@ async def test_arg_from_environ(df_factory):
 
 async def test_unknown_dfly_env(df_factory, export_dfly_password):
     with EnvironCntx(DFLY_abcdef="xyz"):
-        with pytest.raises(DflyStartException):
+        try:
```
Comment (kostasrim) [inline tests/dragonfly/generic_test.py:103]: if you think there is more pythonic way to do this I am open for suggestions @dranikpg

**Review by dranikpg — APPROVED** [review]
👍

**Review by dranikpg — APPROVED** [review]
👍

**Review by dranikpg — APPROVED** [review]
👍

**Comment by dranikpg** [comment]
I'm not a schizophrenic, it's just bad internet on a train 😂

**Comment by romange** [comment]
Maybe I am missing something but it seems obvious why it fails.
1. The process fails to start
2. Its returncode will obviously be non-zero
3. Inside stop() we check for the returncode and throw an exception.

why not avoid introducing a variable and `assert False` by:
```py
with EnvironCntx(DFLY_abcdef="xyz"):
        with pytest.raises(DflyStartException):
            dfly = df_factory.create()
            dfly.start()
       dfly.proc = None  
```

**Review by dranikpg — APPROVED** [review]

**Pull request merged**